rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,221 @@
+
+### Proposal Overview
+
+The entirety of this work lives within the AVM and technically does not change the accessible APIs of the language. It simply fixes the issue that at this time, `@std/datastore` does not share its stored data between instances within the same cluster.
+
+Already within the AVM are two foundational pieces that will be integrated with datastore when in `daemon` mode:
+
+1. LogRendezvousHash to determine what the top `n` nodes are that are assigned the owners of a particular piece of data, with the top node assigned the ability to mutate the data.
+2. The HTTPS and secret token guarded control port, which can be used by the `deploy` service, the nodes of the cluster itself, or the `Anycloud` app to query a particular node for information, such as its status.
+
+We can integrate these two concepts to enable automatic assignment of which nodes will receive the `@std/datastore` key-val pair, and which node will be involved in handling mutations of that data. The control port must have new endpoints that the nodes use to coordinate and pass this data around, and the control port (and these endpoints) can be exposed to the `Anycloud` docker container to enable non-Alan languages to gain the benefit of this automatic, self-healing distributed cache.
+
+This can be broken into multiple tasks:
+
+* Creating the endpoints and the serialization/deserialization of the HandlerMemory necessary
+* Querying LogRendezvousHash for the first node to know which one is "in charge" of this data, or the first `n` nodes if it is itself so it knows where to push the mutation to
+* Fronting the datastore internal data structure with this logic instead of using it directly by the opcodes (when in daemon mode, leaving it as-is or having a "cluster" of one when running locally)
+* Exposing this new behavior into the docker container, and the library work necessary to make this feature usable and natural in other languages.
+* Adding atomic functionality to the datastore API and internals to make distributed mutation (not just key replacement) possible
+
+### Cluster Coordination Work
+
+The new endpoints have some extra logic that they need to tackle, as this is the first time the nodes of the cluster are actually communicating with each other. The LogRendezvousHash module needs to be updated on the *actual* state of the cluster versus the expected state defined in the DNS records: which nodes are actually up or down from its perspective? That determines which nodes are considered for ownership of the datastore data. RendezvousHash (and our own LogRendezvousHash) are stable, in that the top `n` nodes after one of them goes down will still contain `n - 1` of the original set, so not too much data thrashing occurs when a node goes down or comes online.
```

Comment:
We are not proposing that nodes write to the DNS record, right?

Reply:
No, this is their own internal record keeping of the cluster as seen from that nodes' perspective.